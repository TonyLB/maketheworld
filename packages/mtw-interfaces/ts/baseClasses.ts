export class EphemeraError extends Error {
    constructor(message: string) {
        super(message)
        this.name = 'EphemeraException'
    }
}

type EphemeraWrappedId<T extends string> = `${T}#${string}`

export const isEphemeraTaggedId = <G extends string>(tag: G) => (value: string): value is EphemeraWrappedId<G> => {
    const sections = value.split('#')
    if (sections.length > 2) {
        throw new EphemeraError(`Illegal nested EphemeraId: '${value}'`)
    }
    if (sections.length < 2) {
        return false
    }
    return Boolean(sections[0] === tag)
}

export type EphemeraAssetId = EphemeraWrappedId<'ASSET'>
export const isEphemeraAssetId = isEphemeraTaggedId<'ASSET'>('ASSET')

export type EphemeraFeatureId = EphemeraWrappedId<'FEATURE'>
export const isEphemeraFeatureId = isEphemeraTaggedId<'FEATURE'>('FEATURE')

export type EphemeraKnowledgeId = EphemeraWrappedId<'KNOWLEDGE'>
export const isEphemeraKnowledgeId = isEphemeraTaggedId<'KNOWLEDGE'>('KNOWLEDGE')

export type EphemeraRoomId = EphemeraWrappedId<'ROOM'>
export const isEphemeraRoomId = isEphemeraTaggedId<'ROOM'>('ROOM')

export type EphemeraMapId = EphemeraWrappedId<'MAP'>
export const isEphemeraMapId = isEphemeraTaggedId<'MAP'>('MAP')

export type EphemeraCharacterId = EphemeraWrappedId<'CHARACTER'>
export const isEphemeraCharacterId = isEphemeraTaggedId<'CHARACTER'>('CHARACTER')

export type EphemeraActionId = EphemeraWrappedId<'ACTION'>
export const isEphemeraActionId = isEphemeraTaggedId<'ACTION'>('ACTION')

export type EphemeraVariableId = EphemeraWrappedId<'VARIABLE'>
export const isEphemeraVariableId = isEphemeraTaggedId<'VARIABLE'>('VARIABLE')

export type EphemeraComputedId = EphemeraWrappedId<'COMPUTED'>
export const isEphemeraComputedId = isEphemeraTaggedId<'COMPUTED'>('COMPUTED')

export type EphemeraBookmarkId = EphemeraWrappedId<'BOOKMARK'>
export const isEphemeraBookmarkId = isEphemeraTaggedId<'BOOKMARK'>('BOOKMARK')

export type EphemeraMessageId = EphemeraWrappedId<'MESSAGE'>
export const isEphemeraMessageId = isEphemeraTaggedId<'MESSAGE'>('MESSAGE')

export type EphemeraMomentId = EphemeraWrappedId<'MOMENT'>
export const isEphemeraMomentId = isEphemeraTaggedId<'MOMENT'>('MOMENT')

export type EphemeraNotificationId = EphemeraWrappedId<'NOTIFICATION'>
export const isEphemeraNotificationId = isEphemeraTaggedId<'NOTIFICATION'>('NOTIFICATION')

export type EphemeraImageId = EphemeraWrappedId<'IMAGE'>
export const isEphemeraImageId = isEphemeraTaggedId<'IMAGE'>('IMAGE')


export type EphemeraId = EphemeraWrappedId<'ASSET' | 'FEATURE' | 'KNOWLEDGE' | 'ROOM' | 'MAP' | 'CHARACTER' | 'ACTION' | 'VARIABLE' | 'COMPUTED' | 'BOOKMARK' | 'MESSAGE' | 'MOMENT' | 'IMAGE'>
export const isEphemeraId = (value: string): value is EphemeraId => (
    isEphemeraAssetId(value) ||
    isEphemeraFeatureId(value) ||
    isEphemeraKnowledgeId(value) ||
    isEphemeraRoomId(value) ||
    isEphemeraMapId(value) ||
    isEphemeraCharacterId(value) ||
    isEphemeraActionId(value) ||
    isEphemeraVariableId(value) ||
    isEphemeraComputedId(value) ||
    isEphemeraBookmarkId(value) ||
    isEphemeraMessageId(value) ||
    isEphemeraMomentId(value) ||
    isEphemeraImageId(value)
)

export type LegalCharacterColor = 'blue' | 'pink' | 'purple' | 'green' | 'grey'
