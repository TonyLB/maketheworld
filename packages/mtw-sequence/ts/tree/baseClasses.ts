export type GenericTreeNode<N extends {}> = {
    data: N;
    children: GenericTree<N>[];
}

export type GenericTree<N extends {}> = GenericTreeNode<N>[]

