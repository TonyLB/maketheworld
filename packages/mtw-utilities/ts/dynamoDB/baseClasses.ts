import { DynamoDBClient } from "@aws-sdk/client-dynamodb";

//
// TODO: Layer in methods to DBHandler as mixin files
//

export type Constructor<T = {}> = new (...a: any[]) => T

export const remap = <
    K extends string,
    S extends K,
    T extends { [key in K]: any },
    V extends string, // needed to force string literal types for mapping values
    U extends { [P in S]: V }
>(original: T, mapping: U) => {
    const remapped: any = {};

    Object.keys(original).forEach(k => {
        if (mapping[k]) {
            remapped[mapping[k]] = original[k];
        }
        else {
            remapped[k] = original[k]
        }
    })
    return remapped as (Omit<T, keyof U> & {
        // Take all the values in the map, 
        // so given { a: 'Hello', b: 'World', c: '!!!' }  U[keyof U] will produce 'Hello' | 'World' | '!!!'
        [P in keyof U as U[P]]: [T] extends [{ [key in P]: unknown }] ? T[P] : never
    })
}

export type DBHandlerLegalKey = Exclude<string, 'DataCategory'>

export type DBHandlerKey<KInternal extends DBHandlerLegalKey = DBHandlerLegalKey, KeyType extends string = string> = { DataCategory: string } & { [key in KInternal]: KeyType }

export type DBHandlerItem<KInternal extends DBHandlerLegalKey = DBHandlerLegalKey, KeyType extends string = string> = Record<string, any> & { [key in KInternal]: KeyType } & { DataCategory: string }

//
// We export no type restrictions on _internalKeyLabel (since that's a black-box, we shouldn't need to specify it everywhere
// we use a DBHandler, as the constraints will only add value in refactoring DBHandler itself)
//
export class DBHandlerBase<KIncoming extends DBHandlerLegalKey, KeyType extends string = string> {
    _client: InstanceType<typeof DynamoDBClient>;
    _incomingKeyLabel: KIncoming;
    _internalKeyLabel: DBHandlerLegalKey;
    _tableName: string;
    _writeBatchSize?: number;
    _getBatchSize?: number;

    constructor(props: {
        client: InstanceType<typeof DynamoDBClient>;
        tableName: string;
        incomingKeyLabel: KIncoming;
        internalKeyLabel: string;
        options: {
            primaryKeyTypeGuard?: (value: string) => value is KeyType;
            writeBatchSize?: number;
            getBatchSize?: number;
        }
    }, ...rest: any[]) {
        this._client = props.client
        this._tableName = props.tableName
        this._incomingKeyLabel = props.incomingKeyLabel
        this._internalKeyLabel = props.internalKeyLabel
        this._writeBatchSize = props.options.writeBatchSize
        this._getBatchSize = props.options.getBatchSize
    }

    _remapIncomingObject (value: DBHandlerItem<KIncoming, KeyType>): DBHandlerItem
    _remapIncomingObject (value: DBHandlerKey<KIncoming, KeyType>): DBHandlerKey
    _remapIncomingObject (value: DBHandlerItem<KIncoming, KeyType> | DBHandlerKey<KIncoming, KeyType>): DBHandlerItem | DBHandlerKey {
        return remap(value, { [this._incomingKeyLabel]: this._internalKeyLabel } as { [key in KIncoming]: DBHandlerLegalKey }) as (typeof value extends DBHandlerKey<KIncoming, KeyType> ? DBHandlerKey : DBHandlerItem)
    }

    _remapOutgoingObject (value: DBHandlerItem): DBHandlerItem<KIncoming, KeyType>
    _remapOutgoingObject (value: DBHandlerKey): DBHandlerKey<KIncoming, KeyType>
    _remapOutgoingObject (value: DBHandlerItem | DBHandlerKey): DBHandlerItem<KIncoming, KeyType> | DBHandlerKey<KIncoming, KeyType> {
        return remap(value, { [this._internalKeyLabel]: this._incomingKeyLabel } as { [key in DBHandlerLegalKey]: KIncoming }) as (typeof value extends DBHandlerKey ? DBHandlerKey<KIncoming, KeyType> : DBHandlerItem<KIncoming, KeyType>)
    }
}
