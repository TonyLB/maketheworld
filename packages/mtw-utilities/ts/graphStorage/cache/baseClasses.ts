import withGetOperations from "../../dynamoDB/mixins/get";
import withPrimitives from "../../dynamoDB/mixins/primitives";

export type DependencyEdge = {
    EphemeraId: string;
    key?: string;
    assets: string[];
}

export type DependencyNode = {
    EphemeraId: string;
    completeness: 'Partial' | 'Complete';
    connections: DependencyEdge[]
}

export type LegalDependencyTag = 'Asset' | 'Variable' | 'Computed' | 'Room' | 'Feature' | 'Bookmark' | 'Map'
export const isLegalDependencyTag = (tag: string): tag is LegalDependencyTag => (['Asset', 'Variable', 'Computed', 'Room', 'Feature', 'Bookmark', 'Map'].includes(tag))

export type DependencyGraphPut = {
    EphemeraId: string;
    putItem: DependencyEdge;
}

export type DependencyGraphDelete = {
    EphemeraId: string;
    deleteItem: DependencyEdge;
}

export type DependencyGraphAction = DependencyGraphPut | DependencyGraphDelete

export const isDependencyGraphPut = (action: DependencyGraphAction): action is DependencyGraphPut => ('putItem' in action)
export const isDependencyGraphDelete = (action: DependencyGraphAction): action is DependencyGraphDelete => ('deleteItem' in action)

export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>

export type GraphDBHandler<T extends string = string> = InstanceType<ReturnType<ReturnType<typeof withGetOperations<'PrimaryKey', T>>>> & 
    InstanceType<ReturnType<ReturnType<typeof withPrimitives<'PrimaryKey', T>>>>
