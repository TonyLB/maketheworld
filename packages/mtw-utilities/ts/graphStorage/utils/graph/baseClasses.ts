export type GraphEdge <K extends string, E extends Record<string, any>> = E & {
    from: K;
    to: K;
}

export const compareEdges = <K extends string, E extends Record<string, any>>(A: GraphEdge<K, E>, B: GraphEdge<K, E>): number => {
    const fromCompare = A.from.localeCompare(B.from)
    return fromCompare ? fromCompare : A.to.localeCompare(B.to)
}
