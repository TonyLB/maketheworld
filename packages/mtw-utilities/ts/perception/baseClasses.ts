export type ConditionExpression = {
    if: string;
}

export type RoomExit = {
    name: string;
    to: string;             // The asset-local key of the room target
    toEphemeraId: string;   // The globalized key of the room target (e.g. 'VORTEX', not 'ROOM#VORTEX')
}

export type RoomCacheItem = {
    EphemeraId: string;
    name: string[];
    exits: RoomExit[];
}

export type RoomCache = Record<string, RoomCacheItem>
