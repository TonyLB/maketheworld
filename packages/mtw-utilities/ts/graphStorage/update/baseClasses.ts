import { GraphNodeResult } from "../cache/graphNode";
import { Graph } from "../utils/graph";
import withGetOperations from "../../dynamoDB/mixins/get"
import withUpdate from "../../dynamoDB/mixins/update"
import withTransaction from "../../dynamoDB/mixins/transact"
import withBatchWrite from "../../dynamoDB/mixins/batchWrite";

export type GraphOfUpdatesNode = Partial<Omit<GraphNodeResult<string>, 'PrimaryKey'>> & {
    key: string;
    needsForwardUpdate?: boolean;
    needsForwardInvalidate?: boolean;
    forwardInvalidatedAt?: number;
    needsBackUpdate?: boolean;
    needsBackInvalidate?: boolean;
    backInvalidatedAt?: number;
}

export type GraphOfUpdatesEdge = {
    action: 'put' | 'delete';
}

export type GraphStorageDBH = InstanceType<ReturnType<ReturnType<typeof withGetOperations<'PrimaryKey'>>>> &
    InstanceType<ReturnType<ReturnType<typeof withUpdate<'PrimaryKey'>>>> &
    InstanceType<ReturnType<ReturnType<typeof withTransaction<'PrimaryKey'>>>> &
    InstanceType<ReturnType<ReturnType<typeof withBatchWrite<'PrimaryKey'>>>>

export class GraphOfUpdates extends Graph<string, GraphOfUpdatesNode, GraphOfUpdatesEdge> {}

export default GraphOfUpdates
