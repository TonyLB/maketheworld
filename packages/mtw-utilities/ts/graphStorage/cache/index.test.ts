import { CacheBase, GraphDBHandler } from './baseClasses'
import GraphCache from './'
import GraphEdge from './graphEdge'
import GraphNode, { GraphNodeCache, GraphNodeCacheEdge } from './graphNode'
import { marshall } from '@aws-sdk/util-dynamodb'

describe('GraphCache', () => {
    const realDateNow = Date.now.bind(global.Date);

    const dbHandler = {
        getItem: jest.fn(),
        getItems: jest.fn(),
        primitiveUpdate: jest.fn()
    } as unknown as GraphDBHandler
    const internalCache = new ((GraphCache(dbHandler))(GraphEdge(dbHandler)(GraphNode(dbHandler)(CacheBase))))()
    jest.spyOn(internalCache.Edges, 'get')

    const cacheFactory = (entries: string[]): GraphNodeCacheEdge<string>[] => (
        entries.map((key) => ({ from: key.split('::')[0], to: key.split('::')[1], context: '' }))
    )

    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        internalCache.clear()
        const dateNowStub = jest.fn(() => 1000)
        global.Date.now = dateNowStub
    })

    afterEach(() => {
        global.Date.now = realDateNow
    })

    it('should correctly fetch a tree recursively', async () => {
        const nodeGetMock = jest.spyOn(internalCache.Nodes, 'get')
            .mockImplementation((keys) => (
                Promise.resolve(
                    keys.map((key): GraphNodeCache<string> | undefined => ({
                        A: {
                            PrimaryKey: 'A',
                            forward: {
                                edges: [{ target: 'B', context: '' }, { target: 'C', context: '' }],
                            },
                            back: {
                                edges: []
                            }
                        },
                        B: {
                            PrimaryKey: 'B',
                            forward: {
                                edges: [{ target: 'C', context: '' }],
                            },
                            back: {
                                edges: [{ target: 'A', context: '' }]
                            }
                        },
                        C: {
                            PrimaryKey: 'C',
                            forward: {
                                edges: [{ target: 'D', context: '' }, { target: 'E', context: '' }],
                            },
                            back: {
                                edges: [{ target: 'A', context: '' }, { target: 'B', context: '' }]
                            }
                        },
                        D: {
                            PrimaryKey: 'D',
                            forward: {
                                edges: [{ target: 'F', context: '' }],
                            },
                            back: {
                                edges: [{ target: 'C', context: '' }]
                            }
                        },
                        E: {
                            PrimaryKey: 'E',
                            forward: {
                                edges: [{ target: 'F', context: '' }],
                            },
                            back: {
                                edges: [{ target: 'C', context: '' }]
                            }
                        },
                        F: {
                            PrimaryKey: 'F',
                            forward: {
                                edges: [],
                            },
                            back: {
                                edges: [{ target: 'D', context: '' }, { target: 'E', context: '' }]
                            }
                        },
                    }[key])
                ).filter((value): value is GraphNodeCache<string> => (Boolean(value)))
            )
        ))

        jest.spyOn(dbHandler, 'primitiveUpdate').mockResolvedValue({})
        const tree = await internalCache.Graph.get(['A'], 'forward')
        await internalCache.flush()
        expect(nodeGetMock).toHaveBeenCalledTimes(4)
        expect(nodeGetMock).toHaveBeenCalledWith(['A'])
        expect(nodeGetMock).toHaveBeenCalledWith(['B', 'C'])
        expect(nodeGetMock).toHaveBeenCalledWith(['D', 'E'])
        expect(nodeGetMock).toHaveBeenCalledWith(['F'])
        expect(tree.nodes).toEqual({
            A: { key: 'A' },
            B: { key: 'B' },
            C: { key: 'C' },
            D: { key: 'D' },
            E: { key: 'E' },
            F: { key: 'F' }
        })
        expect(tree.edges).toEqual([
            { from: 'A', to: 'B', context: '' },
            { from: 'A', to: 'C', context: '' },
            { from: 'B', to: 'C', context: '' },
            { from: 'C', to: 'D', context: '' },
            { from: 'C', to: 'E', context: '' },
            { from: 'D', to: 'F', context: '' },
            { from: 'E', to: 'F', context: '' }
        ])

        //
        // Should update all caches
        //
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledTimes(6)
        const testArgument = (key: string, cache: string[]) => ({
            Key: { PrimaryKey: key, DataCategory: 'Graph::Forward' },
            UpdateExpression: 'SET cache = :newCache, cachedAt = :moment',
            ConditionExpression: 'attribute_not_exists(cachedAt) OR cachedAt < :moment',
            ExpressionAttributeValues: marshall({
                ':newCache': cache,
                ':moment': 1000
            })
        })
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('A', ['A::B', 'A::C', 'B::C', 'C::D', 'C::E', 'D::F', 'E::F']))
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('B', ['B::C', 'C::D', 'C::E', 'D::F', 'E::F']))
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('C', ['C::D', 'C::E', 'D::F', 'E::F']))
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('D', ['D::F']))
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('E', ['E::F']))
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('F', []))

    })

    it('should correctly fetch a tree with valid caches', async () => {
        const nodeGetMock = jest.spyOn(internalCache.Nodes, 'get')
            .mockImplementation((keys) => (
                Promise.resolve(
                    keys.map((key): GraphNodeCache<string> | undefined => ({
                        A: {
                            PrimaryKey: 'A',
                            forward: {
                                edges: [{ target: 'B', context: '' }, { target: 'C', context: '' }],
                                cache: cacheFactory(['A::B', 'A::C', 'B::C', 'C::D', 'C::E', 'D::F', 'E::F'])
                            },
                            back: {
                                edges: []
                            }
                        },
                        B: {
                            PrimaryKey: 'B',
                            forward: {
                                edges: [{ target: 'C', context: '' }],
                                cache: cacheFactory(['B::C', 'C::D', 'C::E', 'D::F', 'E::F'])
                            },
                            back: {
                                edges: [{ target: 'A', context: '' }]
                            }
                        },
                        C: {
                            PrimaryKey: 'C',
                            forward: {
                                edges: [{ target: 'D', context: '' }, { target: 'E', context: '' }],
                                cache: cacheFactory(['C::D', 'C::E', 'D::F', 'E::F'])
                            },
                            back: {
                                edges: [{ target: 'A', context: '' }, { target: 'B', context: '' }]
                            }
                        },
                        D: {
                            PrimaryKey: 'D',
                            forward: {
                                edges: [{ target: 'F', context: '' }],
                                cache: cacheFactory(['D::F'])
                            },
                            back: {
                                edges: [{ target: 'C', context: '' }]
                            }
                        },
                        E: {
                            PrimaryKey: 'E',
                            forward: {
                                edges: [{ target: 'F', context: '' }],
                                cache: cacheFactory(['E::F'])
                            },
                            back: {
                                edges: [{ target: 'C', context: '' }]
                            }
                        },
                        F: {
                            PrimaryKey: 'F',
                            forward: {
                                edges: [],
                                cache: []
                            },
                            back: {
                                edges: [{ target: 'D', context: '' }, { target: 'E', context: '' }]
                            }
                        },
                    }[key])
                ).filter((value): value is GraphNodeCache<string> => (Boolean(value)))
            )
        ))

        jest.spyOn(dbHandler, 'primitiveUpdate').mockResolvedValue({})
        const tree = await internalCache.Graph.get(['A'], 'forward')
        await internalCache.flush()
        expect(nodeGetMock).toHaveBeenCalledTimes(2)
        expect(nodeGetMock).toHaveBeenCalledWith(['A'])
        expect(nodeGetMock).toHaveBeenCalledWith(['B', 'C', 'D', 'E', 'F'])
        expect(Object.keys(tree.nodes)).toEqual(['A', 'B', 'C', 'D', 'E', 'F'])
        expect(tree.edges).toEqual([
            { from: 'A', to: 'B', context: '' },
            { from: 'A', to: 'C', context: '' },
            { from: 'B', to: 'C', context: '' },
            { from: 'C', to: 'D', context: '' },
            { from: 'C', to: 'E', context: '' },
            { from: 'D', to: 'F', context: '' },
            { from: 'E', to: 'F', context: '' }
        ])

        //
        // Does not need to update any caches
        //
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledTimes(0)

    })

    it('should correct a tree with inconsistent caches', async () => {
        const nodeGetMock = jest.spyOn(internalCache.Nodes, 'get')
            .mockImplementation((keys) => (
                Promise.resolve(
                    keys.map((key): GraphNodeCache<string> | undefined => ({
                        A: {
                            PrimaryKey: 'A',
                            forward: {
                                edges: [{ target: 'B', context: '' }, { target: 'C', context: '' }],
                                cache: cacheFactory(['A::B', 'A::C', 'B::C', 'C::D', 'D::F'])
                            },
                            back: {
                                edges: []
                            }
                        },
                        B: {
                            PrimaryKey: 'B',
                            forward: {
                                edges: [{ target: 'C', context: '' }],
                                cache: cacheFactory(['B::C', 'C::D', 'C::E', 'D::F', 'E::F'])
                            },
                            back: {
                                edges: [{ target: 'A', context: '' }]
                            }
                        },
                        C: {
                            PrimaryKey: 'C',
                            forward: {
                                edges: [{ target: 'D', context: '' }, { target: 'E', context: '' }],
                                cache: cacheFactory(['C::D', 'C::E', 'D::F', 'E::F'])
                            },
                            back: {
                                edges: [{ target: 'A', context: '' }, { target: 'B', context: '' }]
                            }
                        },
                        D: {
                            PrimaryKey: 'D',
                            forward: {
                                edges: [{ target: 'F', context: '' }],
                                cache: cacheFactory(['D::F'])
                            },
                            back: {
                                edges: [{ target: 'C', context: '' }]
                            }
                        },
                        E: {
                            PrimaryKey: 'E',
                            forward: {
                                edges: [{ target: 'F', context: '' }],
                                cache: cacheFactory(['E::F'])
                            },
                            back: {
                                edges: [{ target: 'C', context: '' }]
                            }
                        },
                        F: {
                            PrimaryKey: 'F',
                            forward: {
                                edges: [],
                                cache: []
                            },
                            back: {
                                edges: [{ target: 'D', context: '' }, { target: 'E', context: '' }]
                            }
                        },
                    }[key])
                ).filter((value): value is GraphNodeCache<string> => (Boolean(value)))
            )
        ))

        jest.spyOn(dbHandler, 'primitiveUpdate').mockResolvedValue({})
        const tree = await internalCache.Graph.get(['A'], 'forward')
        await internalCache.flush()
        expect(nodeGetMock).toHaveBeenCalledTimes(3)
        expect(nodeGetMock).toHaveBeenCalledWith(['A'])
        expect(nodeGetMock).toHaveBeenCalledWith(['B', 'C', 'D', 'F'])
        expect(nodeGetMock).toHaveBeenCalledWith(['E'])
        expect(Object.keys(tree.nodes)).toEqual(['A', 'B', 'C', 'D', 'F', 'E'])
        expect(tree.edges).toEqual([
            { from: 'A', to: 'B', context: '' },
            { from: 'A', to: 'C', context: '' },
            { from: 'B', to: 'C', context: '' },
            { from: 'C', to: 'D', context: '' },
            { from: 'C', to: 'E', context: '' },
            { from: 'D', to: 'F', context: '' },
            { from: 'E', to: 'F', context: '' }
        ])

        //
        // Updates only the inconsistent cache
        //
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledTimes(1)
        const testArgument = (key: string, cache: string[]) => ({
            Key: { PrimaryKey: key, DataCategory: 'Graph::Forward' },
            UpdateExpression: 'SET cache = :newCache, cachedAt = :moment',
            ConditionExpression: 'attribute_not_exists(cachedAt) OR cachedAt < :moment',
            ExpressionAttributeValues: marshall({
                ':newCache': cache,
                ':moment': 1000
            })
        })
        expect(dbHandler.primitiveUpdate).toHaveBeenCalledWith(testArgument('A', ['A::B', 'A::C', 'B::C', 'C::D', 'C::E', 'D::F', 'E::F']))

    })

})