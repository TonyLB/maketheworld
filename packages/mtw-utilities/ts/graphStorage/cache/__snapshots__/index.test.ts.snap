// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegacyGraphCache delete should correctly delete a connection 1`] = `
Array [
  Object {
    "EphemeraId": "VARIABLE#testOne",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
      },
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`LegacyGraphCache delete should decrement connection assets when redundancies exist 1`] = `
Array [
  Object {
    "EphemeraId": "VARIABLE#testOne",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`LegacyGraphCache generationOrder should correctly calculate generation order for a complex graph 1`] = `
Array [
  Array [
    "VARIABLE#One",
  ],
  Array [
    "COMPUTED#Two",
    "COMPUTED#Six",
  ],
  Array [
    "COMPUTED#Four",
  ],
  Array [
    "COMPUTED#Seven",
  ],
]
`;

exports[`LegacyGraphCache get should correctly fetch a tree 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
    "tag": "Computed",
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Complete",
    "connections": Array [],
    "tag": "Computed",
  },
]
`;

exports[`LegacyGraphCache getPartial should correctly decipher a partial tree 1`] = `
Array [
  Object {
    "EphemeraId": "VARIABLE#testOne",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
      },
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`LegacyGraphCache put should backpopulate antidependency links 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "assets": Array [],
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testFour",
        "assets": Array [
          "base",
        ],
        "key": "testFour",
      },
    ],
    "tag": "Computed",
  },
  Object {
    "EphemeraId": "COMPUTED#testFour",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`LegacyGraphCache put should correctly add a complete tree 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testFour",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
        "key": "testFour",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`LegacyGraphCache put should correctly add a single graph item 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testFour",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
        "key": "testFour",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;
